### PR TITLE
Alternate fair/unfair lock modes in the connection pool to improve throughput

### PR DESCRIPTION
**Pull request by stIncMale**
Acquiring locks unfairly in `OpenConcurrencyLimiter.releasePermit`
revealed a bug in `OpenConcurrencyLimiter.acquirePermitOrGetAvailableOpenedConnection`,
which manifested itself by failing the `pool-checkout-minPoolSize-connection-maxConnecting.json`
test. A fix had to be implemented together with the performance-related change.

Another tiny change is moving `pool.release(openConnection)` out of the guarded section
in the `OpenConcurrencyLimiter.tryHandOverOrRelease`: it does not have to be there,
and when not only fair locking is used, this change has a significant positive measurable effect
on the connection cycle (check out/check in) throughput.

For more details see this [report](https://jira.mongodb.org/browse/JAVA-4435?focusedCommentId=4311277&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-4311277).

JAVA-4435

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java [hunk 1]
```diff
@@ -992,12 +994,10 @@ private PooledConnection acquirePermitOrGetAvailableOpenedConnection(final boole
...
-                while (permits == 0) {
-                    stateAndGeneration.throwIfClosedOrPaused();
-                    availableConnection = tryGetAvailable ? tryGetAvailableConnection() : null;
-                    if (availableConnection != null) {
-                        break;
-                    }
+                while (permits == 0
+                        // the absence of short-circuiting is of importance
+                        & (availableConnection = tryGetAvailable ? tryGetAvailableConnection() : null) == null
+                        & !stateAndGeneration.throwIfClosedOrPaused()) {
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java:1000]: I know we don't usually use assignment expressions as expressions, but this is the situation when making an exception is worth it. The bug mentioned in the description of the PR was caused by how the `while` condition was specified. It's a condition that the thread waits for, so it has to be checked before calling `awaitNanos`, and after returning from this method. Having the full condition specified in the `while (...)` part not only achieves this, but also achieves this without code duplication.

Originally, the condition was split: part was inside `while (...)`, and part was inside the `while` body. The second part should have been duplicated after the `while` loop in order to make sure that the whole condition is checked again after returning from the last `awaitNanos` call, but it wasn't there.

I could have fixed the code by duplicating the second part of the condition
```java
stateAndGeneration.throwIfClosedOrPaused();
availableConnection = tryGetAvailable ? tryGetAvailableConnection() : null;
```
after the `while` loop, but it only makes the code more complex and leaves opportunities for future mistakes if the condition is ever changed. I think that putting the whole condition inside `while (...)` is very much the right thing to do, even if it looks unconventional compared to the rest of the driver code.

**Review by jyemin — CHANGES_REQUESTED** [review]

**Comment by stIncMale** [comment]
Closing. Two different Jira tickets will be created instead of JAVA-4435, and two different PRs will be created instead of this PR.

File: driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java [hunk 2]
```diff
@@ -992,12 +994,10 @@ private PooledConnection acquirePermitOrGetAvailableOpenedConnection(final boole
                     expressedDesireToGetAvailableConnection = true;
                 }
                 long remainingNanos = timeout.remainingOrInfinite(NANOSECONDS);
-                while (permits == 0) {
-                    stateAndGeneration.throwIfClosedOrPaused();
-                    availableConnection = tryGetAvailable ? tryGetAvailableConnection() : null;
-                    if (availableConnection != null) {
-                        break;
-                    }
+                while (permits == 0
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java:997]: Oh, thank you! Yes, the code was incorrect, I changed the order in the new PR (https://github.com/mongodb/mongo-java-driver/pull/858) to make sure no exceptions are thrown if `availableConnection` becomes non-null.

File: driver-core/src/main/com/mongodb/internal/connection/ConcurrentPool.java [hunk 3]
```diff
@@ -373,20 +399,24 @@ boolean acquirePermitImmediateUnfair() {
...
          */
-        boolean acquirePermitFair(final long timeout, final TimeUnit unit) throws MongoInterruptedException {
+        boolean acquirePermit(final long timeout, final TimeUnit unit) throws MongoInterruptedException {
             long remainingNanos = unit.toNanos(timeout);
-            try {
-                // preserve the eager InterruptedException behavior of `Semaphore.tryAcquire(long, TimeUnit)`
-                lock.writeLock().lockInterruptibly();
-            } catch (InterruptedException e) {
-                throw new MongoInterruptedException(null, e);
+            if (waitersEstimate.get() == 0) {
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/connection/ConcurrentPool.java:409]: We discussed this in Slack, and I left note about this in the code comment: https://github.com/mongodb/mongo-java-driver/pull/862/files#diff-2319c4409b718a7bd5bdd204abb211abd311f5b1630f73ca5deafe933a36ada9R364-R368. This way no future maintainer will be tempted to remove `waitersEstimate` and start relying on `permits`.

File: driver-core/src/main/com/mongodb/internal/connection/ConcurrentPool.java [hunk 4]
```diff
@@ -497,4 +532,37 @@ boolean closed() {
...
+    }
+
+    static void lockUnfair(final ReentrantLock lock) {
+        if (!lock.tryLock()) {
+            // `ReentrantLock.tryLock` is always unfair, but if it does not succeed, fall back to a potentially fair method
+            lock.lock();
+        }
+    }
+
+    private static void lockUnfair(final ReentrantReadWriteLock.WriteLock lock) {
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/connection/ConcurrentPool.java:556]: `Lock` is unaware of the concept of fairness. Thus, calling `Lock.tryLock` and saying that it behaves unfairly does not make much sense. On the other hand, the documentation of `ReentrantReadWriteLock.WriteLock.tryLock` and `ReentrantLock.tryLock` are explicit about those methods behaving unfair.

**Pull request closed without merging**